rfc/008-platform-support.md

```diff
@@ -0,0 +1,111 @@
+# RFC 008: Platform Support
+
+## Objective
+
+Implement support for diverse platforms in Contrast, ensuring that:
+
+1. Users can mix-and-match features throughout all platforms supported by Contrast
+2. Code with a platform dependency can be deduplicated
+3. New platforms and platform features can be added with low complexity (that's `O(1)`, at best)
+
+## Problem statement
+
+Contrast currently distinguishes between platforms based on their distinct features.
+Taking the `AKS-CLH-SNP` platform as an example, this serialized platform string conveys
+the following information:
+
+- Kubernetes engine: `AKS`
+  - (Implicit: Kata upstream / Microsoft fork)
+- Hypervisor: `CLH`
+- Hardware platform: `SNP`
+
+The user picks such a platform through its serialized string (for example `AKS-CLH-SNP`), and the
+code then implements actions based upon that, such as when writing platform-specific Kata
+configuration values in the [node-installer].
+
+With new features being added to Contrast, such as GPU support, more information needs
+to be contained in such a platform description.
+
+This comes at the cost of growing complexity of conditional statements
+depending on the platform throughout the implementation code, as the
+current implementation doesn't distinguish between platform features
+(for example which hypervisor is used) but only between full platform combinations.
+
+[node-installer]: https://github.com/edgelesssys/contrast/blob/3f39682ea9a383b2557923e257bd065e461b8ee6/nodeinstaller/internal/constants/constants.go#L48
+
+## Proposal
+
+### Structured Representation
+
+Throughout the implementation code, platforms shouldn't be represented as _individual combinations_,
+but rather through a generic, structured type that contains information about their _features_.
+
+So `AKS-CLH-SNP` could become something like this:
+
+```go
+type Platform struct {
+    KubernetesDistribution KubernetesDistribution
```

Comment:
We must think about what we really need to know. Like, is the distribution relevant, or do we really only need the paths and the systemd unit to restart? Otherwise we are just pushing the problem down a layer.